contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,846 @@
+# StatefulSet Updates
+
+**Author**: kow3ns@
+
+**Status**: Proposal
+
+## Abstract
+Currently (as of Kubernetes 1.6), `.Spec.Replicas`, and 
+`.Spec.Template.Containers` are the only mutable fields of the 
+StatefulSet API object. Updating `.Spec.Replicas` will scale the number of Pods 
+in the StatefulSet. Updating `.Spec.Template.Containers` causes all subsequently 
+created Pods to have the specified containers. In order to cause the 
+StatefulSet controller to apply its updated `.Spec`, users must manually delete 
+each Pod. This manual method of applying updates is error prone. The 
+implementation of this proposal will add the capability to perform ordered,
+automated, sequential updates. 
+
+## Affected Components
+1. API Server
+1. Kubectl
+1. StatefulSet Controller
+1. StatefulSetSpec API object
+1. StatefulSetStatus API object
+
+## Use Cases
+Upon implementation, this design will support the following in scope use cases, 
+and it will not rule out the future implementation of the out of scope use 
+cases.
+
+### In Scope
+- As the administrator of a stateful application, in order to vertically scale 
+my application, I want to update resource limits or requested resources.
+- As the administrator of a stateful application, in order to deploy critical 
+security updates, break fix patches, and feature releases, I want to update 
+container images.
+- As the administrator of a stateful application, in order to update my 
+application's configuration, I want to update environment variables, container 
+entry point commands or parameters, or configuration files.
+- As the administrator of the logging and monitoring infrastructure for my 
+organization, in order to add logging and monitoring side cars, I want to patch
+containers to add images.
+
+### Out of Scope
+- As the administrator of a stateful application, in order to increase the 
+applications storage capacity, I want to update PersistentVolumes.
+- As the administrator of a stateful application, in order to update the 
+network configuration of the application, I want to update Services and 
+container ports in a consistent way.
+- As the administrator of a stateful application, when I scale my application 
+horizontally, I want associated PodDistruptionBudgets to be adjusted to 
+compensate for the application's scaling.
+
+## Assumptions
+ - StatefulSet update must support singleton StatefulSets. However, an update in
+ this case will cause a temporary outage. This is acceptable as a single 
+ process application is, by definition, not highly available.
+ - Disruption in Kubernetes is controlled by PodDistruptionBudgets. As 
+ StatefulSet updates progress one Pod at a time, and only occur when all 
+ other Pods have a Status of Running and a Ready Condition, they can not 
+ violate reasonable PodDisrutptionBugdets.
```

Comment:
The point of the above is that, given the implementations limitations, the controller will not violate any reasonable PodDisruptionBudget by triggering an update. If we ever want to update more than one Pod at a time than we would have to create an eviction to be safe.